packages/vaadin-themable-mixin/src/css-utils.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * @license
+ * Copyright (c) 2021 - 2025 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+
+/**
+ * Injects the given stylesheet into the shadow root of the component
+ * through the adoptedStyleSheets API. The stylesheet is always added
+ * at the beginning of the adoptedStyleSheets array.
+ *
+ * @param {HTMLElement} component
+ * @param {CSSStyleSheet} styleSheet
+ */
+export function injectStyleSheet(component, styleSheet) {
+  // Remove the old stylesheet to avoid duplicates
+  const adoptedStyleSheets = component.shadowRoot.adoptedStyleSheets.filter(
+    (s) => s !== component.__cssInjectorStyleSheet,
+  );
+
+  // Add the new stylesheet at the beginning of the adoptedStyleSheets array
+  component.shadowRoot.adoptedStyleSheets = [styleSheet, ...adoptedStyleSheets];
```

Comment:
This relies on the assumption that the new base styles will use `@layer` but maybe we could improve this:

- Update `finalizeStyles()` method in `ThemableMixin` to store styles from `static get styles()` separately
- Have some unified helper that would then apply styles in the right order when updating instance styles:
  - first, styles from `static get styles()`
  - next, styles injected by `CSSInjector`
  - finally, styles added with `registerStyles()`

If this sounds reasonable, I will prototype this in a separate branch.

Reply:
UPD: implemented this logic in a new commit and updated tests accordingly.